### PR TITLE
[daikin] Support the BRP1B61 (aka Airbase) controller

### PR DESCRIPTION
**Pull request by psmedley**
Signed-off-by: Paul Smedley <paul@smedley.id.au>

Add support for the BRP1B61 controller which uses a similar (but slightly) different API.

### CODE REVIEW

**Comment by psmedley** [comment]
Note: this was previously planned to be a separate binding in https://github.com/openhab/openhab2-addons/pull/5777

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java [hunk 1]
```diff
@@ -48,6 +56,7 @@
  * Handles communicating with a Daikin air conditioning unit.
  *
  * @author Tim Waterhouse - Initial Contribution
+ * @author Paul Smedley <paul@smedley.id.au> - Modifications to support Airbase Controllers
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: You might consider creating a new class `AirbaseAcUnitHandler` extending the class `DaikinAcUnitHandler` and put all Airbase specific code into that class.  For specific code make or add in the `DaikinAcUnitHandler` methods protected and override them in the new class. For example with `pollStatus`.  You simply override this method and put the content of the method `pollAirbaseStatus` in the overridden method. That way you don't have to do checks on thing id and keep the code cleaner. In the Factory class you can simply instantiate with the specific handler.

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinHandlerFactory.java [hunk 2]
```diff
@@ -27,8 +27,10 @@
  * handlers.
  *
  * @author Tim Waterhouse <tim@timwaterhouse.com> - Initial contribution
+ * @author Paul Smedley <paul@smedley.id.au> - Modifications to support Airbase Controllers
+
  */
-@Component(service = ThingHandlerFactory.class, configurationPid = "binding.daikin")
+@Component(service = ThingHandlerFactory.class, immediate = true, configurationPid = "binding.daikin")
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinHandlerFactory.java]: Thanks for the comments @Hilbrand I'll review over the weekend and hopefully get time to commit updates. I had thought of an array for the zones but wasn't sure at the time how to implement - my java is pretty rusty :)

[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinHandlerFactory.java]: Fixed in f239b2af75

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java [hunk 3]
```diff
@@ -59,6 +84,47 @@ public SensorInfo getSensorInfo() throws DaikinCommunicationException {
         return SensorInfo.parse(response);
     }
 
+    //Daikin Airbase API
+    public AirbaseControlInfo AirbaseGetControlInfo() throws DaikinCommunicationException {
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java]: Fixed in f239b2af75

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseBasicInfo.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Holds information from the basic_info call.
+ *
+ * @author Paul Smedley - Initial contribution
+ *
+ */
+public class AirbaseBasicInfo {
+    private static Logger LOGGER = LoggerFactory.getLogger(AirbaseBasicInfo.class);
+
+    public String SSID;
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseBasicInfo.java]: Fixed in f239b2af75

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseBasicInfo.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                }).collect(Collectors.toMap(x -> x[0], x -> x[1]));
+
+        AirbaseBasicInfo info = new AirbaseBasicInfo();
+        info.SSID = responseMap.get("ssid");
+        return info;
+    }
+
+    public Map<String, String> getParamString() {
+        Map<String, String> params = new HashMap<>();
+        params.put("ssid", SSID.toString());
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseBasicInfo.java]: Fixed in f239b2af75

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    String[] keyValue = kv.split("=");
+                    String key = keyValue[0];
+                    String value = keyValue.length > 1 ? keyValue[1] : "";
+                    return new String[] { key, value };
+                }).collect(Collectors.toMap(x -> x[0], x -> x[1]));
+
+        AirbaseZoneInfo info = new AirbaseZoneInfo();
+        info.zonenames = responseMap.get("zone_name");
+        String zoneinfo = responseMap.get("zone_onoff");
+        StringTokenizer zones = new StringTokenizer(zoneinfo, "%3b");
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java]: Fixed in https://github.com/openhab/openhab2-addons/pull/5910/commits/dda01670971e406ce73aa102c167f58f07a4f5e6

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ * Holds information from the basic_info call.
+ *
+ * @author Paul Smedley - Initial contribution
+ *
+ */
+public class AirbaseZoneInfo {
+    private static Logger LOGGER = LoggerFactory.getLogger(AirbaseZoneInfo.class);
+
+    public String zonenames;
+    public boolean zone1 = false;
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java]: Fixed in https://github.com/openhab/openhab2-addons/pull/5910/commits/dda01670971e406ce73aa102c167f58f07a4f5e6

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java [hunk 8]
```diff
@@ -101,7 +118,60 @@ private void handleCommandInternal(ChannelUID channelUID, Command command) throw
...
+
+                    if (thingTypeUID.equals(DaikinBindingConstants.THING_TYPE_AC_UNIT))
+                        changeMode(Mode.valueOf(((StringType) command).toString()));
+                    else
+                        changeAirbaseMode(AirbaseMode.valueOf(((StringType) command).toString()));
+                    return;
+                }
+                break;
+            /* additional controls for Daikin Airbase */
+            case DaikinBindingConstants.CHANNEL_AIRBASE_AC_ZONE1:
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: I've attempted to simplify this in 8af64248de - I'm sure the case code could be simplified more but not sure how to do that yet..

[the remark this answers is not in the document]
Comment (Hilbrand) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: You can simply the code as follows. Remove the zone case statements, add a new zone constant `CHANNEL_AIRBASE_AC_ZONE` and add the following code after the switch block (The `channelId` variable can be move to the beginning of the method and also used in the switch):

```
String channelId = channelUID.getId();
if (channelId.startsWith(DaikinBindingConstants.CHANNEL_AIRBASE_AC_ZONE)) {
    int zoneNumber = Integer.parseInt(channelUID.getId().substring(4));
    if (command instanceof OnOffType) {
         changeZone(zoneNumber, command == OnOffType.ON);
         return;
    }
}
```

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java [hunk 9]
```diff
@@ -191,16 +268,57 @@ private void pollStatus() throws IOException {
...
+        SensorInfo sensorInfo = webTargets.getAirbaseSensorInfo();
+        if (sensorInfo != null) {
+            updateTemperatureChannel(DaikinBindingConstants.CHANNEL_INDOOR_TEMP, sensorInfo.indoortemp);
+
+            updateTemperatureChannel(DaikinBindingConstants.CHANNEL_OUTDOOR_TEMP, sensorInfo.outdoortemp);
+        }
+        AirbaseModelInfo modelInfo = webTargets.getAirbaseModelInfo();
+        AirbaseZoneInfo zoneInfo = webTargets.getAirbaseZoneInfo();
+        if (zoneInfo != null) {
+            updateState(DaikinBindingConstants.CHANNEL_AIRBASE_AC_ZONE1, zoneInfo.zone[1] ? OnOffType.ON : OnOffType.OFF);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: You can replace all these state changes with a single statement:
```
IntStream.range(0, zoneInfo.zone.length).forEach(
    idx -> updateState(DaikinBindingConstants.CHANNEL_AIRBASE_AC_ZONE + idx, OnOffType.from(zoneInfo.zone[idx])));
```

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java [hunk 10]
```diff
@@ -42,5 +46,17 @@
     public static final String CHANNEL_AC_FAN_DIR = "fandir";
     public static final String CHANNEL_HUMIDITY = "humidity";
 
-    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(THING_TYPE_AC_UNIT);
+    // additional channels for Airbase Controller
+    public static final String CHANNEL_AIRBASE_AC_FAN_SPEED = "airbasefanspeed";
+    public static final String CHANNEL_AIRBASE_AC_ZONE1 = "zone1";
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java]: You should be able to remove all these individual zone channels with a generic: `CHANNEL_AIRBASE_AC_ZONE = "zone";` . See my other comments.

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: Thanks - addressed in 3d2e3c53f7

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java [hunk 11]
```diff
@@ -59,6 +84,47 @@ public SensorInfo getSensorInfo() throws DaikinCommunicationException {
...
+    }
+
+    public AirbaseZoneInfo getAirbaseZoneInfo() throws DaikinCommunicationException {
+        String response = invoke(getAirbaseZoneInfoUri);
+        return AirbaseZoneInfo.parse(response);
+    }
+
+    public void setAirbaseZoneInfo(AirbaseZoneInfo zoneinfo, AirbaseModelInfo modelinfo) throws DaikinCommunicationException {
+        int count=0;
+        count = (zoneinfo.zone[1] ? 1 : 0) + (zoneinfo.zone[2] ? 1 : 0) + (zoneinfo.zone[3] ? 1 : 0) + (zoneinfo.zone[4] ? 1 : 0) + (zoneinfo.zone[5] ? 1 : 0) + (zoneinfo.zone[6] ? 1 : 0) + (zoneinfo.zone[7] ? 1 : 0) + (zoneinfo.zone[8] ? 1 : 0) + modelinfo.commonzone;
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java [hunk 12]
```diff
@@ -59,6 +84,47 @@ public SensorInfo getSensorInfo() throws DaikinCommunicationException {
...
+        logger.debug("Number of open zones: \"{}\"", count);
+
+        Map<String, String> queryParams = zoneinfo.getParamString();
+        if (count >= 1) invoke(setAirbaseZoneInfoUri, queryParams);
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinWebTargets.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+public class AirbaseEnums {
+    public enum AirbaseMode {
+        UNKNOWN(-1),
+        FAN(0),
+        HEAT(1),
+        COLD(2),
+        DRY(7),
+        AUTO(3);
+
+        private static Logger LOGGER = LoggerFactory.getLogger(AirbaseMode.class);
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java]: Fixed in 4cafb310da - I also updated this in some of the original source files that you hadn't commented on

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public enum AirbaseFanMovement {
+        UNKNOWN(-1),
+        STOPPED(0),
+        VERTICAL(1),
+        HORIZONTAL(2),
+        VERTICAL_AND_HORIZONTAL(3);
+
+        private static Logger LOGGER = LoggerFactory.getLogger(AirbaseFanMovement.class);
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+            .append("%3b")
+            .append(zone[5] ? "1" : "0")
+            .append("%3b")
+            .append(zone[6] ? "1" : "0")
+            .append("%3b")
+            .append(zone[7] ? "1" : "0")
+            .append("%3b")
+            .append(zone[8] ? "1" : "0");
+        params.put("zone_name", zonenames);
+        params.put("zone_onoff", onoffstring.toString());
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseZoneInfo.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java [hunk 16]
```diff
@@ -232,15 +360,37 @@ private void changeMode(Mode mode) throws DaikinCommunicationException {
...
         info.fanMovement = fanDir;
         webTargets.setControlInfo(info);
     }
+
+    // Zones only supported on Airbase, so don't need to detect what kind of thing we have
+    private void changeZone(int zone, boolean command) throws DaikinCommunicationException {
+        AirbaseZoneInfo info = webTargets.getAirbaseZoneInfo();
+        AirbaseModelInfo modelInfo = webTargets.getAirbaseModelInfo();
+        info.zone[zone] = command;
+        if (modelInfo.zonespresent >=zone) webTargets.setAirbaseZoneInfo(info, modelInfo);
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java (hunk 9 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java (hunk 10 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+
+    public enum AirbaseFanSpeed {
+        AUTO("A"),
+        LEVEL_1("1"),
+        LEVEL_2("3"),
+        LEVEL_3("5");
+
+        private static Logger LOGGER = LoggerFactory.getLogger(AirbaseFanSpeed.class);
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/api/airbase/AirbaseEnums.java]: Fixed in 4cafb310da

File: bundles/org.openhab.binding.daikin/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 18]
```diff
@@ -32,6 +32,42 @@
...
+			<channel id="zone3" typeId="airbase-acunit-zone3" />
+			<channel id="zone4" typeId="airbase-acunit-zone4" />
+			<channel id="zone5" typeId="airbase-acunit-zone5" />
+			<channel id="zone6" typeId="airbase-acunit-zone6" />
+			<channel id="zone7" typeId="airbase-acunit-zone7" />
+			<channel id="zone8" typeId="airbase-acunit-zone8" />
+
+		</channels>
+
+		<config-description>
```
[the remark this answers is not in the document]
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/resources/ESH-INF/thing/thing-types.xml]: Fixed in 2b4d2cd3b9

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java (hunk 1 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: I see this one is flagged as 'might consider'. Can this one be a continuous improvement item for later? I have other things on the roadmap once the basic airbase changes are in git

Comment (Hilbrand) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: Yes I think that is ok.

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java (hunk 10 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/DaikinBindingConstants.java]: Can you remove the constants in comment. There not needed any more I assume.

File: bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java (hunk 1 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikin/src/main/java/org/openhab/binding/daikin/internal/handler/DaikinAcUnitHandler.java]: Thanks. I'll keep it in mind for future PRs

**Comment by psmedley** [comment]
> With one minor comment about commented dead code I think this pr is ready to be merged when that is fixed.

The commented dead code is fixed in e26a688ee6

**Pull request merged**